### DISCUSSION TITLE
Node-dependent caching of queries with aggressive caching

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#13970] [Q&A] [answered]
We observe node-dependent caching with aggressive caching on a three-node cluster with round-robin load balancing.

When a query is executed in a session where we set the document store to aggressively cache, it is executed once for each node in the cluster. After it's been executed on each node once, additional requests are not sent to the server at all.

We use per-session aggressive caching like this:
```
using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromMinutes(1), AggressiveCacheMode.DoNotTrackChanges))
{
      // query
}
```

Is this intended behavior? Can it be customized if it's intended? Although it's possible to have the indexes in different states on the individual nodes, I would expect the cached value to be used even when the new session was assigned to a different node.

**Comment by ayende** [comment, marked as answer]
That is expected, yes. The cache itself is done at the level of the HTTP requests, so different hostnames are threated differently.

**Reply by krivulcik** [reply]
While this is understandable on the technical level, it was surprising.

Also, conceptually, I would expect document-store-specific caching to be pretty reasonable.

Would it be prohibitively difficult to consider the cache relevant for all nodes in a cluster? It would improve the overall performance and behavior of the system as a whole.

**Reply by ayende** [reply]
Yes, it would be quite hard to do. In practice, you should not generally hit this, mind.

We tend to be stick on a single node by default (unless you define `ReadLoadBalancing`), which means that you'll use the same node all the time unless there is a failover.

The cache itself is implemented at a very low level (HTTP requests) and is global, so all operations are taking part of that.

For documents, on a theoretical basis, you'll get the right value across nodes, but that isn't something that we actually guarantee. 

For indexes, for etag computation, we are using values that are local to each node, so the same query on different nodes will have different etags, even if they have the same actual source data.

Trying to ensure consistent distributed etag generation across the cluster is a... challenge, and not something that we intend to do.

**Reply by krivulcik** [reply]
Yes, this makes sense. Thanks for the confirmation. I thought it was probably done too low in the architecture to allow caching by the whole document store easily.

We use read balance behavior (round-robin), so we usually hit the cluster three times for each cached value. While already improving the performance and behavior of the system, the three requests are often a significant fraction of all requests in our case, so it's a bit unfortunate that this can't be improved, but it's not a deal-breaker.

**Reply by ayende** [reply]
See if you can use the context load balancing instead, that would tie each scenario to particular node, giving you better cache locality.
https://ravendb.net/docs/article-page/5.0/java/client-api/session/configuration/use-session-context-for-load-balancing

**Reply by krivulcik** [reply]
Yes, that's what we ended up doing, seems to work well. Thanks for the assistance.
